Report stale status

brb opened the pull request:
This PR:

- Introduces `pkg/status` which adds a mechanism for running status probes concurrently. 
- Moves status retrieval of each `daemon` subsystem into separates probe, so a deadlocking/stale status update can be reported.

Example of a stale report:
```bash
    $ cilium status
    KVStore:                Failure   Err: No response from kvstore probe within 15 seconds
    ContainerRuntime:       Ok        docker daemon: OK
    Kubernetes:             Failure   No response from kubernetes probe within 15 seconds
    Kubernetes APIs:        [""]
    Cilium:                 Warning   Stale status data
    Stale status:           "kubernetes" since 2018-11-21T11:11:45.164Z, "kvstore" since 2018-11-21T11:09:43.133Z
    NodeMonitor:            Disabled
    Cilium health daemon:   Ok
    IPv4 address pool:      260/65535 allocated
    IPv6 address pool:      3/65535 allocated
    Proxy Status:           OK, ip 10.0.0.1, port-range 10000-20000
```

Notes:
- I considered adding a smoke test, but in the end I didn't write such, as w/o adding some flags to `daemon` the test would be relatively complex and flaky.
- There is a `FIXME` for adding the locks status. It's going to be done in a separate PR, as this PR already became fairly big.
- We should probably change the order when making the final decision for `Cilium` status, as IMO any failure in `KvStore`/`ContainerRuntime`/`Kubernetes` is more important than the staleness reporting. It can be done in a separate PR.
- `cilium debuginfo` won't wait for all probes to return, so some status fields might be missing.
- Currently no probe is using `context.Context`, so the param is redundant and can be removed. However, it would be a good idea to start using it.

Joint work with @tgraf and @raybejjani.

Fixes: #5674

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/6271)
<!-- Reviewable:end -->

brb committed:
status: add pkg for running status probes concurrently

pkg/status/status.go introduces a mechanism for running status probes
concurrently.

The main use-case is to detect stale statuses of the daemon subsystems.

Joint work with @tgraf and @raybejjani.

Signed-off-by: Martynas Pumputis <martynas@covalent.io>

brb committed:
api: add the field Stale to StatusResponse

The field is a map which key is a stale probe name and value is
a timestamp when the probe was started.

Joint work with @tgraf and @raybejjani.

Signed-off-by: Martynas Pumputis <martynas@covalent.io>

brb committed:
daemon: get status of each subsystem concurrently

Move status collection of each subsystem into probes handled by
the collector from pkg/status.

Each probe is run concurrently, so a deadlocking/stale probe a) does not
stop other probes, b) can be reported.

Joint work with @tgraf and @raybejjani.

Signed-off-by: Martynas Pumputis <martynas@covalent.io>

brb committed:
client: report stale statuses (probes) with status cmd

Example of such report:

```
$ cilium status
KVStore:                Failure   Err: No response from kvstore probe within 15 seconds
ContainerRuntime:       Ok        docker daemon: OK
Kubernetes:             Failure   No response from kubernetes probe within 15 seconds
Kubernetes APIs:        [""]
Cilium:                 Warning   Stale status data
Stale status:           "kubernetes" since 2018-11-21T11:11:45.164Z, "kvstore" since 2018-11-21T11:09:43.133Z
NodeMonitor:            Disabled
Cilium health daemon:   Ok
IPv4 address pool:      260/65535 allocated
IPv6 address pool:      3/65535 allocated
Proxy Status:           OK, ip 10.0.0.1, port-range 10000-20000
```

Joint work with @tgraf and @raybejjani.

Fixes #5674.

Signed-off-by: Martynas Pumputis <martynas@covalent.io>

vadorovsky commented on pkg/status/status_test.go:
```diff
@@ -0,0 +1,174 @@
+	c.Assert(testutils.WaitUntil(func() bool {
+		return atomic.LoadUint64(&ok) >= 2
+	}, 1*time.Second), IsNil)
+	c.Assert(len(collector.GetStaleProbes()), Equals, 1)
```
```
c.Assert(collector.GetStaleProbes(), HasLen, 1)
```

vadorovsky commented on pkg/status/status_test.go:
```diff
@@ -0,0 +1,174 @@
+	c.Assert(testutils.WaitUntil(func() bool {
+		return atomic.LoadUint64(&ok) >= 1
+	}, 1*time.Second), IsNil)
+	c.Assert(len(collector.GetStaleProbes()), Equals, 1)
```
```
c.Assert(collector.GetStaleProbes(), HasLen, 1)
```

vadorovsky commented on pkg/status/status_test.go:
```diff
@@ -0,0 +1,174 @@
+	c.Assert(testutils.WaitUntil(func() bool {
+		return atomic.LoadUint64(&ok) >= 3 && atomic.LoadUint64(&errors) >= 3
+	}, 1*time.Second), IsNil)
+	c.Assert(len(collector.GetStaleProbes()), Equals, 0)
```
```
c.Assert(collector.GetStaleProbes(), HasLen, 0)
```

vadorovsky commented on pkg/status/status_test.go:
```diff
@@ -0,0 +1,174 @@
+	c.Assert(testutils.WaitUntil(func() bool {
+		return atomic.LoadUint64(&timeout) == 2 && atomic.LoadUint64(&ok) > 0
+	}, 1*time.Second), IsNil)
+	c.Assert(len(collector.GetStaleProbes()), Equals, 0)
```
```
c.Assert(collector.GetStaleProbes(), HasLen, 0)
```

vadorovsky commented on daemon/status.go:
```diff
@@ -85,90 +81,275 @@ func (d *Daemon) getNodeStatus() *models.ClusterStatus {
+	defer d.statusCollectMutex.RUnlock()
 
-	sr.Kubernetes = d.getK8sStatus()
+	// TODO(brb) we should do a deep copy instead, as statusResponse contains some references
```
And why can't we do a deep copy now?

brb replied:
I couldn't find how to generate with `go-swagger` a method for the deep copy. Any idea?

brb replied:
I added a dependency for doing a deep copy in this particular case, which is IMHO a bit an overkill. Alternatively, I could add a method for the deep copy manually, or just do `Unmarshal(Marshal(...))`.

vadorovsky commented on daemon/status.go:
```diff
@@ -85,90 +81,275 @@ func (d *Daemon) getNodeStatus() *models.ClusterStatus {
+					msg = fmt.Sprintf("Err: %s", status.Err)
+				case ok:
+					msg = fmt.Sprintf("%s", info)
+				}
```
I don't see the `!ok` case being handled here. Am I missing something? Or can we just ignore situations when `Data` has a different type?

brb replied:
I was reluctant to handle the `!ok` case, as it can only happen when someone changes the `"kvstore"`'s `Probe` method in a way that it returns a non-string type, and it would require adding more boilerplate (if we do here, then probably we would need to add to other probes as well).

vadorovsky replied:
OK, makes sense, thanks for explanation. I'm fine with not handling `!ok` case then.

vadorovsky commented on pkg/status/status.go:
```diff
@@ -0,0 +1,237 @@
+)
+
+const (
+	defaultInterval         = 5 * time.Second
```
We usually store default consts in `pkg/default`.

ianvernon replied:
If the constants are localized to a package, then I think it makes more sense to keep them within that package. If a constant is utilized by multiple packages, then it should be in `pkg/default`.

brb replied:
I'm fine with both approaches. Maybe keeping in `pkg/default` makes a default more visible.

vadorovsky reviewed (changes requested):
Overall looks good, great job! I have few questions/comments though.

brb committed:
{daemon,status,defaults}: address PR comments

This commit is going to be squashed after all reviews.

Signed-off-by: Martynas Pumputis <martynas@covalent.io>

brb committed:
deps: add github.com/mohae/deepcopy

It's going to be used by daemon/status.go for making
a deep copy of models.StatusResponse.

Signed-off-by: Martynas Pumputis <martynas@covalent.io>

brb committed:
daemon: do deep copy of model.StatusResponse

This commit is going to be squashed after all reviews.

Signed-off-by: Martynas Pumputis <martynas@covalent.io>

ianvernon reviewed (approved):
This is great! Documentation is A+ and made it easy to understand what the code was doing, especially in the core logic of the status probing. Great job!

tgraf reviewed (approved):
Awesome work!
